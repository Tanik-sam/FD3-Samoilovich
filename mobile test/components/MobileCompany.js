import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';
import MobileClientCard from './MobileClientCard';
import {clientEvents} from './events';







class MobileCompany extends React.PureComponent {

  static propTypes = {
    
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        surname: PropTypes.string.isRequired,
        nameCl: PropTypes.string.isRequired,
        patronymic: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    
    clients: this.props.clients,
    surname: this.props.surname,
    nameCl:this.props.nameCl,
    patronymic: this.props.patronymic,
    status:  this.props.balance>0?true:false,
    balance:this.props.balance,
    clientsEdt: this.props.clients.slice(),
    filtered:0, //0--все, 1 -активные, 2- заблокированные
    selectedClientId: null,
    cardShown: [],
    cardMode:1,
    editMode:1,
  };


  

  componentDidMount = () => {
    clientEvents.addListener('ClientClicked',this.clientSelected);
    clientEvents.addListener('ClientDelete',this.clientDelete);
    clientEvents.addListener('ClientEdit',this.clientEdit);
    clientEvents.addListener('ClientSave',this.clientSave);
    clientEvents.addListener('ClientDefault',this.clientDefault);
    clientEvents.addListener('ButtonEnbled',this.buttonEnbled)
    
  };

  componentWillUnmount = () => {
    clientEvents.removeListener('ClientClicked',this.clientSelected);
    clientEvents.addListener('ClientDelete',this.clientDelete);
    clientEvents.removeListener('ClientEdit',this.clientEdit);
    clientEvents.removeListener('ClientSave',this.clientSave);
    clientEvents.removeListener('ClientDefault',this.clientDefault);
    clientEvents.addListener('ButtonEnbled',this.buttonEnbled)
  };
  buttonEnbled=()=>{
    this.setState({editMode:1})
    }
  clientDelete=(clDl)=>{
    function fff(v,i,a){return v.id!=clDl}
    let kk=[...this.state.clientsEdt]
    let k=kk.filter(fff);
     this.setState({cardShown:[]})
     this.setState({clientsEdt:k})
     
  }



  clientSave=(clSn)=>{
  
    
  let k=[...this.state.clientsEdt]
  let changed=false
  let changedCl=[]
    for (let i=0;i<k.length;i++){
      if ((k[i].id==clSn.id)&&((k[i].surname!=clSn.surname)||(k[i].nameCl!=clSn.nameCl)||(k[i].patronymic!=clSn.patronymic)||(k[i].balance!=clSn.balance))){
        //k[i]=clSn;
        let newClientData={...k[i]}
        newClientData.surname=clSn.surname
        newClientData.nameCl=clSn.nameCl
        newClientData.patronymic=clSn.patronymic
        newClientData.balance=clSn.balance
        changed=true
        k[i]=newClientData
        changedCl=[newClientData]
        
      }
      this.setState({cardMode:1})
      clientEvents.emit('ButtonEnbled');
    }
    if (changed==true){
      this.setState({clientsEdt:k})
      this.setState ({cardShown:changedCl}) 

    }
  
  }

  clientDefault=(clD)=>{ this.setState({cardMode:1})}

  clientEdit = (edtCdVl) => {
    this.setState({cardMode:2})
    this.setState({editMode:0})
   }
  addClient=()=>{
    var ll=[...this.state.clientsEdt];
    var llll=ll.map(v=>v.id)
    let maxId = Math.max.apply(Math, llll) 
    var nn={id:(maxId+1),surname:"",nameCl:"",patronymic:"",status:false,balance:0}
    
    ll=[...ll,nn]
    this.setState({clientsEdt:ll})
    let lastCl=[];
    lastCl=[...lastCl,nn]
    this.setState ({cardShown:lastCl})
    this.setState({cardMode:2})
  }
  clientSelected=(clId)=>{
    
    this.setState({selectedClientId: clId} ); 
    function ffff(v,i,a){return v.id==clId }
    let cl=[...this.state.clientsEdt]; 
    let cll=cl.filter(ffff); 
    this.setState ({cardShown:cll})
   
    }

allClients=()=>{
  this.setState({filtered:0}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
activeClients=()=>{
  this.setState({filtered:1}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
blockedClients=()=>{
  this.setState({filtered:2}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
  render() {

    var cG=[];
    for (var a=0; a<this.props.columnName.length; a++ ) {
      var columnN=this.props.columnName[a];
      let cGs = <th className='ColumnN' key={columnN.code}>  {columnN.text}</th>
      cG.push(cGs);
    }

    console.log("MobileCompany render");
    console.log(this.state.clientsEdt);
    var clientFiltered=this.state.clientsEdt
    if (this.state.filtered==1){let clientActive=[...this.state.clientsEdt]; 
      function f1(v,i,a){return v.balance>0 }
      clientFiltered=clientActive.filter(f1);}
      else {
      if (this.state.filtered==2){let clientBlocked=[...this.state.clientsEdt]; 
        function f2(v,i,a){return v.balance<=0 }
        clientFiltered=clientBlocked.filter(f2);}
     }
    var clientsCode=clientFiltered.map(client =>
      <MobileClient key={client.id} info={client} selectedClientId={this.state.selectedClientId} />
    );
    

    if (this.state.cardShown!=[]) {var clientSelected=this.state.cardShown.map(v=>
      <MobileClientCard  info={v} cardMode={this.state.cardMode}
      nameRow={this.props.columnName} selectedClientId={this.selectedClientId}/>)}
    return (
      <div className='MobileCompany'>
        <input className='notTable' type="button" value="Все" onClick={this.allClients}  />
        <input className='notTable' type="button" value="Активные" onClick={this.activeClients}  />
        <input className='notTable' type="button" value="Заблокированные" onClick={this.blockedClients}  />
        <hr/>
        <div className='MobileCompanyClients'>
          <table>
            <tbody>
        <tr className="MobileClient">{cG}</tr> 
          {clientsCode}
          </tbody>
          </table>
        </div>
        <input className="inTable" type="button" value="Добавить клиента" onClick={this.addClient} disabled={this.state.editMode==0} />
        {clientSelected}
      </div>
    )
    ;

  }

}

export default MobileCompany;
