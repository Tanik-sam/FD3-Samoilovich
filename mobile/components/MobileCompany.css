.MobileCompany
{
  border: solid #667 1px;
  border-radius: 30px;
  padding: 3px 5px;
  background-color:  #D0ECF4;
}

.MobileCompanyName
{
  padding: 7px 7px;
  font-size: large;
  font-family: Verdana, Geneva, Tahoma, sans-serif; 
  font-weight: bold;
  color:#667;
}

.MobileCompanyClients
{
  padding: 10px 0px;
  
  
}

th {
  color:#667;
  width:200px;
  text-align:center;
  padding: 8px 8px;
  font-size: large;
  font-family: Verdana, Geneva, Tahoma, sans-serif; 
  
}
tr
{
  background: linear-gradient(45deg, #b1e092 #c97aee);
  
}
hr{
  border-color:  rgb(166, 166, 190);
  border-width:thin
  
}