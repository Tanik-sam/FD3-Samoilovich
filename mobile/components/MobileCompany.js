import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';
import MobileClientCard from './MobileClientCard';
import {clientEvents} from './events';






class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        surname: PropTypes.string.isRequired,
        nameCl: PropTypes.string.isRequired,
        patronymic: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients,
    surname: this.props.surname,
    nameCl:this.props.nameCl,
    patronymic: this.props.patronymic,
    status:  this.props.balance>0?true:false,
    balance:this.props.balance,
    clientsEdt: this.props.clients.slice(),
    filtered:0, //0--все, 1 -активные, 2- заблокированные
    selectedClientId: null,
    cardShown: [],
    cardMode:1,
  };

  setName1 = () => {
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  
  

  componentDidMount = () => {
    clientEvents.addListener('ClientClicked',this.clientSelected);
    clientEvents.addListener('ClientDelete',this.clientDelete);
    clientEvents.addListener('ClientEdit',this.clientEdit);
    clientEvents.addListener('ClientSave',this.clientSave);
    clientEvents.addListener('ClientDefault',this.clientDefault);
    
  };

  componentWillUnmount = () => {
    clientEvents.removeListener('ClientClicked',this.clientSelected);
    clientEvents.addListener('ClientDelete',this.clientDelete);
    clientEvents.removeListener('ClientEdit',this.clientEdit);
    clientEvents.removeListener('ClientSave',this.clientSave);
    clientEvents.removeListener('ClientDefault',this.clientDefault);
  };

  clientDelete=(clDl)=>{
    function fff(v,i,a){return v.id!=clDl}
    let kk=[...this.state.clientsEdt]
    let k=kk.filter(fff);
     this.setState({cardShown:[]})
     this.setState({clientsEdt:k})
     
  }



  clientSave=(clSn)=>{
  let k=[...this.state.clientsEdt]
  let changed=false
    for (let i=0;i<k.length;i++){
      if ((k[i].id==clSn.id)&&((k[i].surname!=clSn.surname)||(k[i].surname!=clSn.surname)||(k[i].surname!=clSn.surname)||(k[i].surname!=clSn.surname))){
        k[i]=clSn;
        changed=true
       console.log('нашли неравные')
       }
    }
    if (changed==true){
      this.setState({clientsEdt:k})
      
    }
  
  }

  clientDefault=(clD)=>{ this.setState({cardMode:1})}

  clientEdit = (edtCdVl) => {
    this.setState({cardMode:2})
   }
  addClient=()=>{
    
    var lll=this.state.clientsEdt.length+101
    var nn={id:lll,surname:"",nameCl:"",patronymic:"",status:true,balance:0}
    var ll=[...this.state.clientsEdt];
    //ll.push(nn)
    ll=[...ll,nn]
    this.setState({clientsEdt:ll})
    
    this.clientSelected(lll)
    this.setState({cardMode:2})
  }
  clientSelected=(clId)=>{
    this.setState({selectedClientId: clId} ); 
    function ffff(v,i,a){return v.id==clId }
    let cl=[...this.state.clientsEdt]; 
    let cll=cl.filter(ffff); 
    this.setState ({cardShown:cll})
   
    }

allClients=()=>{
  this.setState({filtered:0}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
activeClients=()=>{
  this.setState({filtered:1}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
blockedClients=()=>{
  this.setState({filtered:2}) //0--все, 1 -активные, 2- заблокированные
  this.setState ({cardShown:[]})
}
  render() {

    var cG=[];
    for (var a=0; a<this.props.columnName.length; a++ ) {
      var columnN=this.props.columnName[a];
      let cGs = <th className='ColumnN' key={columnN.code}>  {columnN.text}</th>
      cG.push(cGs);
    }

    console.log("MobileCompany render");
    console.log(this.state.clientsEdt);
    var clientFiltered=this.state.clientsEdt
    if (this.state.filtered==1){let clientActive=[...this.state.clientsEdt]; 
      function f1(v,i,a){return v.balance>0 }
      clientFiltered=clientActive.filter(f1);}
      else {
      if (this.state.filtered==2){let clientBlocked=[...this.state.clientsEdt]; 
        function f2(v,i,a){return v.balance<=0 }
        clientFiltered=clientBlocked.filter(f2);}
     }
    var clientsCode=clientFiltered.map(client =>
      <MobileClient key={client.id} info={client} selectedClientId={this.state.selectedClientId} />
    );
    

    if (this.state.cardShown!=[]) {var clientSelected=this.state.cardShown.map(v=>
      <MobileClientCard  key={v.id} info={v} cardMode={this.state.cardMode}
      nameRow={this.props.columnName} selectedClientId={this.selectedClientId}/>)}
    return (
      <div className='MobileCompany'>
        <input className='notTable' type="button" value="=МТС" onClick={this.setName1} />
        <input className='notTable' type="button" value="=Velcom" onClick={this.setName2} />
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <hr/>
        <input className='notTable' type="button" value="Все" onClick={this.allClients}  />
        <input className='notTable' type="button" value="Активные" onClick={this.activeClients}  />
        <input className='notTable' type="button" value="Заблокированные" onClick={this.blockedClients}  />
        <hr/>
        <div className='MobileCompanyClients'>
          <table>
            <tbody>
        <tr className="MobileClient">{cG}</tr> 
          {clientsCode}
          </tbody>
          </table>
        </div>
        <input className="inTable" type="button" value="Добавить клиента" onClick={this.addClient} />
        {clientSelected}
      </div>
    )
    ;

  }

}

export default MobileCompany;
