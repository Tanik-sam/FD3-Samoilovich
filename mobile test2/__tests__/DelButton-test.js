"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import MobileCompany from '../components/MobileCompany';
test('Проверка кнопок "Удалить"', () => {
  
  // создаём тестовую версию компонента
   const component = renderer.create(
    <MobileCompany clients={[ 
      {id:101, surname:"Иванов",nameCl:"Иван",patronymic:"Иванович", balance:2}, 
      {id:102, surname:"Сидоров ",nameCl:"Сидор",patronymic:"Сидорович ", balance:250}, 
      {id:103, surname:"Петров",nameCl:"Петр",patronymic:"Петрович", balance:180}, 
      {id:104, surname:"Григорьев",nameCl:"Григорий",patronymic:"Григорьевич", balance:0}, 
     ]} columnName={[ 
      {text:'Фамилия',code:0}, 
      {text:'Имя',code:1}, 
      {text:'Отчество',code:2},
      {text:'Баланс',code:3},
      {text:'Статус',code:4},
      {text:'Редактировать',code:5},
      {text:'Удалить',code:6},
    ]}/>
    
  );
  let clients=[ 
    {id:101, surname:"Иванов",nameCl:"Иван",patronymic:"Иванович", balance:2}, 
    {id:102, surname:"Сидоров ",nameCl:"Сидор",patronymic:"Сидорович ", balance:250}, 
    {id:103, surname:"Петров",nameCl:"Петр",patronymic:"Петрович", balance:180}, 
    {id:104, surname:"Григорьев",nameCl:"Григорий",patronymic:"Григорьевич", balance:0}, 
   ]

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  for (let i=0;i<clients.length;i++){
  // найдём в вёрстке компонента саму кнопку
  console.log (clients.length)
  let dataDel='del'+clients[i].id
  console.log (i)
  console.log (dataDel)
  const buttonElem = component.root.find(el => el.type=='input' && el.props.data==dataDel ); 
  console.log (buttonElem)

  buttonElem.props.onClick();
  }

  // получаем уже изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // "нажмём" кнопку ещё раз
  buttonElem.props.onClick();
  
  // и получаем окончательный снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  
});
