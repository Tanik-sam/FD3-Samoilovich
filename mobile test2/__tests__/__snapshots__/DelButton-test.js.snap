// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Проверка кнопок "Удалить" 1`] = `
<div
  className="MobileCompany"
>
  <input
    className="notTable"
    data="all"
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    className="notTable"
    data="active"
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    className="notTable"
    data="blocked"
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <div
    className="MobileCompanyClients"
  >
    <table>
      <tbody>
        <tr
          className="MobileClient"
        >
          <th
            className="ColumnN"
          >
              
            Фамилия
          </th>
          <th
            className="ColumnN"
          >
              
            Имя
          </th>
          <th
            className="ColumnN"
          >
              
            Отчество
          </th>
          <th
            className="ColumnN"
          >
              
            Баланс
          </th>
          <th
            className="ColumnN"
          >
              
            Статус
          </th>
          <th
            className="ColumnN"
          >
              
            Редактировать
          </th>
          <th
            className="ColumnN"
          >
              
            Удалить
          </th>
        </tr>
        <tr
          className="MobileClient"
          onClick={[Function]}
        >
          <td
            className="StatusActive"
          >
            Иванов
          </td>
          <td
            className="StatusActive"
          >
            Иван
          </td>
          <td
            className="StatusActive"
          >
            Иванович
          </td>
          <td
            className="StatusActive"
          >
            2
          </td>
          <td
            className="StatusActive"
          >
            active
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              data="del101"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
             
          </td>
        </tr>
        <tr
          className="MobileClient"
          onClick={[Function]}
        >
          <td
            className="StatusActive"
          >
            Сидоров 
          </td>
          <td
            className="StatusActive"
          >
            Сидор
          </td>
          <td
            className="StatusActive"
          >
            Сидорович 
          </td>
          <td
            className="StatusActive"
          >
            250
          </td>
          <td
            className="StatusActive"
          >
            active
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              data="del102"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
             
          </td>
        </tr>
        <tr
          className="MobileClient"
          onClick={[Function]}
        >
          <td
            className="StatusActive"
          >
            Петров
          </td>
          <td
            className="StatusActive"
          >
            Петр
          </td>
          <td
            className="StatusActive"
          >
            Петрович
          </td>
          <td
            className="StatusActive"
          >
            180
          </td>
          <td
            className="StatusActive"
          >
            active
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              data="del103"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
             
          </td>
        </tr>
        <tr
          className="MobileClient"
          onClick={[Function]}
        >
          <td
            className="StatusActive"
          >
            Григорьев
          </td>
          <td
            className="StatusActive"
          >
            Григорий
          </td>
          <td
            className="StatusActive"
          >
            Григорьевич
          </td>
          <td
            className="StatusActive"
          >
            0
          </td>
          <td
            className="StatusBlocked"
          >
            blocked
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Редактировать"
            />
          </td>
          <td
            className="StatusActive"
          >
            <input
              className="inTable"
              data="del104"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="Удалить"
            />
             
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <input
    className="inTable"
    data="add"
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;
